File: bennu-core/src/main/java/pt/ist/bennu/core/domain/Theme.java
/* 
 * @(#)Theme.java 
 * 
 * Copyright 2009 Instituto Superior Tecnico 
 * Founding Authors: João Figueiredo, Luis Cruz, Paulo Abrantes, Susana Fernandes 
 *  
 *      https://fenix-ashes.ist.utl.pt/ 
 *  
 *   This file is part of the Bennu Web Application Infrastructure. 
 * 
 *   The Bennu Web Application Infrastructure is free software: you can 
 *   redistribute it and/or modify it under the terms of the GNU Lesser General 
 *   Public License as published by the Free Software Foundation, either version  
 *   3 of the License, or (at your option) any later version. 
 * 
 *   Bennu is distributed in the hope that it will be useful, 
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of 
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
 *   GNU Lesser General Public License for more details. 
 * 
 *   You should have received a copy of the GNU Lesser General Public License 
 *   along with Bennu. If not, see <http://www.gnu.org/licenses/>. 
 *  
 */
package pt.ist.bennu.core.domain;

import pt.ist.fenixframework.Atomic;

/**
 * 
 * @author Luis Cruz
 * @author Paulo Abrantes
 * 
 */
public class Theme extends Theme_Base {

    public Theme(String name, ThemeType type) {
        super();
        setMyOrg(MyOrg.getInstance());
        setName(name);
        setType(type);
    }

    public Theme(String name, String description, ThemeType type) {
        this(name, type);
        setDescription(description);
    }

    public Theme(String name, String description, ThemeType type, String screenshotFileName) {
        this(name, description, type);
        setScreenshotFileName(screenshotFileName);
    }

    public void delete() {
        setMyOrg(null);
        for (VirtualHost virtualHost : getVirtualHostsSet()) {
            virtualHost.setTheme(null);
        }
        deleteDomainObject();
    }

    public static Theme getThemeByName(String name) {
        for (Theme theme : MyOrg.getInstance().getThemesSet()) {
            if (theme.getName().equals(name)) {
                return theme;
            }
        }
        return null;
    }

    public static boolean isThemeAvailable(String name) {
        return getThemeByName(name) != null;
    }

    @Override
    public String getScreenshotFileName() {
        String screenshotName = super.getScreenshotFileName();
        return screenshotName != null ? screenshotName : getName() + ".jpg";
    }

    @Atomic
    public static Theme createTheme(String themeName, String description, ThemeType type, String screenshotFileName) {
        return new Theme(themeName, description, type, screenshotFileName);
    }

    @Atomic
    public static void deleteTheme(Theme theme) {
        theme.delete();
    }
    @Deprecated
    public java.util.Set<pt.ist.bennu.core.domain.VirtualHost> getVirtualHosts() {
        return getVirtualHostsSet();
    }

}

File: bennu-toolkit/src/main/webapp/bennu-toolkit/js/theme.js
/*
 * theme.js
 * 
 * Copyright (c) 2014, Instituto Superior Técnico. All rights reserved.
 * 
 * This file is part of Bennu Toolkit.
 * 
 * Bennu Toolkit is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 * 
 * Bennu Toolkit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along with Bennu Toolkit. If not, see
 * <http://www.gnu.org/licenses/>.
 */

(function () {

    Bennu.theme = Bennu.theme || {};
    var cache = {};
    function hackColor(cls){
    	if (!(cls in cache)){
	    	var a = $("<button style='display:none;' class='btn btn-" + cls + "'></button>"); 
	    	$(document.body).append(a); 
	    	var color = a.css("backgroundColor");
	    	a.remove();
	    	cache[cls] = color;
	    	return color;
    	}else{
    		return cache[cls];
    	}
    }

    function hackParagraph(){
        if (!("paragraph" in cache)){
            var a = $("<p style='display:none;'></p>"); 
            $(document.body).append(a); 
            var font = a.css("fontFamily");
            a.remove();
            cache["paragraph"] = font;
            return font;
        }else{
            return cache["paragraph"];
        }
    }

    function hackHeading(){
        if (!("heading" in cache)){
            var a = $("<h1 style='display:none;'></h1>"); 
            $(document.body).append(a); 
            var font = a.css("fontFamily");
            a.remove();
            cache["heading"] = font;
            return font;
        }else{
            return cache["heading"];
        }
    }

    Bennu.theme.hFontFamily = function(){
        return hackHeading();
    };

    Bennu.theme.pFontFamily = function(){
        return hackParagraph();
    };

    Bennu.theme.primary = function(){
    	return hackColor("primary");
    };

    Bennu.theme.success = function(){
    	return hackColor("success");
    };

    Bennu.theme.info = function(){
    	return hackColor("info");
    };

    Bennu.theme.warning = function(){
    	return hackColor("warning");
    };

    Bennu.theme.danger = function(){
    	return hackColor("danger");
    };

})();